components/raftstore-v2/src/operation/command/admin/leader_transfer.rs

```diff
@@ -0,0 +1,419 @@
+        // log is always its current term. Not just set term because raft library
+        // forbids setting it for MsgTransferLeader messages.
+        msg.set_log_term(self.term());
+        self.raft_group_mut().raft.msgs.push(msg);
+        true
+    }
+
+    pub fn on_transfer_leader_msg<T>(
+        &mut self,
+        ctx: &mut StoreContext<EK, ER, T>,
+        msg: &eraftpb::Message,
+        peer_disk_usage: DiskUsage,
+    ) {
+        // log_term is set by original leader, represents the term last log is written
+        // in, which should be equal to the original leader's term.
+        if msg.get_log_term() != self.term() {
+            return;
+        }
+
+        if !self.is_leader() {
+            self.execute_transfer_leader(ctx, msg.get_from(), peer_disk_usage, false);
+        } else {
+            let from = match self.peer_from_cache(msg.get_from()) {
+                Some(p) => p,
+                None => return,
+            };
+            match self.ready_to_transfer_leader(ctx, msg.get_index(), &from) {
+                Some(reason) => {
+                    info!(
+                        self.logger,
+                        "reject to transfer leader";
+                        "to" => ?from,
+                        "reason" => reason,
+                        "index" => msg.get_index(),
+                        "last_index" => self.storage().last_index().unwrap_or_default(),
+                    );
+                }
+                None => {
+                    self.propose_pending_writes(ctx);
+                    if self.propose_locks_before_transfer_leader(ctx, msg) {
+                        // If some pessimistic locks are just proposed, we propose another
+                        // TransferLeader command instead of transferring leader immediately.
+                        info!(
+                            self.logger,
+                            "propose transfer leader command";
+                            "to" => ?from,
+                        );
+                        let mut cmd =
+                            new_admin_request(self.region().get_id(), self.peer().clone());
+                        cmd.mut_header()
+                            .set_region_epoch(self.region().get_region_epoch().clone());
+                        // Set this flag to propose this command like a normal proposal.
+                        cmd.mut_header()
+                            .set_flags(WriteBatchFlags::TRANSFER_LEADER_PROPOSAL.bits());
+                        cmd.mut_admin_request()
+                            .set_cmd_type(AdminCmdType::TransferLeader);
+                        cmd.mut_admin_request().mut_transfer_leader().set_peer(from);
+                        if let (PeerMsg::RaftCommand(req), sub) = PeerMsg::raft_command(cmd) {
+                            self.on_admin_command(ctx, req.request, req.ch);
+                        } else {
+                            unreachable!();
```

Comment:
i don't think we need unreachable!() here when on_admin_command failed. Just log error instead.

Reply:
unreachable here means PeerMsg::raft_command should only return PeerMsg::RaftCommand type.